apollo-router/src/plugins/headers.rs

```diff
@@ -345,25 +345,36 @@ impl<S> HeadersService<S> {
                         }
                     }
                     Insert::FromBody(from_body) => {
-                        let output = from_body
-                            .path
-                            .execute(req.supergraph_request.body())
-                            .ok()
-                            .flatten();
-                        if let Some(val) = output {
-                            let header_value = if let Value::String(val_str) = val {
-                                val_str
-                            } else {
-                                val.to_string()
-                            };
-                            match HeaderValue::from_str(&header_value) {
-                                Ok(header_value) => {
+                        if body_to_value.is_none() {
+                            body_to_value =
+                                serde_json_bytes::value::to_value(req.supergraph_request.body())
+                                    .ok();
+                        }
+
+                        if let Some(body_to_value) = &body_to_value {
+                            let output = from_body.path.find(body_to_value);
+                            if let serde_json_bytes::Value::Null = output {
+                                if let Some(default_val) = &from_body.default {
                                     req.subgraph_request
                                         .headers_mut()
-                                        .insert(&from_body.name, header_value);
+                                        .insert(&from_body.name, default_val.clone());
                                 }
-                                Err(err) => {
-                                    tracing::error!("cannot convert from the body into a header value for header name '{}': {:?}", from_body.name, err);
+                            } else {
+                                let header_value =
+                                    if let serde_json_bytes::Value::String(val_str) = output {
+                                        val_str.as_str().to_string()
+                                    } else {
+                                        output.to_string()
+                                    };
+                                match HeaderValue::from_str(&header_value) {
+                                    Ok(header_value) => {
+                                        req.subgraph_request
+                                            .headers_mut()
+                                            .insert(&from_body.name, header_value);
+                                    }
+                                    Err(err) => {
+                                        tracing::error!("cannot convert from the body into a header value for header name '{}': {:?}", from_body.name, err);
```

Comment:
NIT, I prefer having the errors separately for json parsing purposes. So:
```suggestion
                                        let header_name = from_body.name;
                                        tracing::error!(header_name, err?, "cannot convert from the body into a header value for header name");
```
I think that syntax is correct, I haven't typed it for a while...

Reply:
done in https://github.com/apollographql/router/pull/6621/commits/3032728cffe78efd9ece24f7ab2c33259ae44827